This is a classic regression problem where we predict an insurance rate based on a person's health and other information.

## Part 1

Load and inspect the data

In [32]:
# Imports
import pandas as pd

df = pd.read_csv('../data/insurance.csv')

## ai note and issue. this was the first time i cloned and pushed a homework assignment both on my home computer w a mac in jupyter notebook, and i opened it and worked on it at my work laptop which is in VSCODE. there was some funkiness when i used vscode where it was taking forever to run the basic df descriptive columns, so i used the github help ai add on at work to troubleshoot why it was so slow. it kept trying to fix my kernel and then at work it wasnt picking up the definition of root mean squared error and kept having me define it. i just ran the final one in jupyter notebook back on my personal computer and everything seems to be working again, so i am pushing this copy 

In [33]:
df.columns

Index(['age', 'sex', 'bmi', 'children', 'smoker', 'region', 'charges'], dtype='str')

In [34]:
print(df.shape)

(1338, 7)


In [35]:
df.head()


,age,sex,bmi,children,smoker,region,charges
0,19,female,27.900,0,yes,southwest,16884.92400
1,18,male,33.770,1,no,southeast,1725.55230
2,28,male,33.000,3,no,southeast,4449.46200
3,33,male,22.705,0,no,northwest,21984.47061
4,32,male,28.880,0,no,northwest,3866.85520


In [36]:
df.info()

<class 'pandas.DataFrame'>
RangeIndex: 1338 entries, 0 to 1337
Data columns (total 7 columns):
 #   Column    Non-Null Count  Dtype  
---  ------    --------------  -----  
 0   age       1338 non-null   int64  
 1   sex       1338 non-null   str    
 2   bmi       1338 non-null   float64
 3   children  1338 non-null   int64  
 4   smoker    1338 non-null   str    
 5   region    1338 non-null   str    
 6   charges   1338 non-null   float64
dtypes: float64(2), int64(2), str(3)
memory usage: 73.3 KB


---

## Part 2.

Simple Linear Regression.

* Regress one numeric feature on the target variable `charges`
* Set up your `X` and `y` first.
* Do a train test split, default test size is fine. MAKE SURE TO SET A RANDOM STATE.
* Fit the model.
* Interpret the coefficient.

In [37]:
from sklearn.linear_model import LinearRegression
from sklearn.model_selection import train_test_split
from sklearn.metrics import mean_absolute_error, root_mean_squared_error, r2_score



In [38]:
# Create X and y
X = df[['bmi']]
y = df['charges']



In [39]:
# Train test split
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)



In [40]:
# create a new instance of LinearRegression
lr = LinearRegression()

# fit it on your training data.
lr.fit(X_train, y_train)

# make predictions store as y_pred
y_pred = lr.predict(X_test)



In [41]:
# Evaluate your model with MAE, RMSE, R2 Score
print('MAE: ', mean_absolute_error(y_test, y_pred))
print('RMSE:', root_mean_squared_error(y_test, y_pred))
print('R2:  ', r2_score(y_test, y_pred))



MAE:  9456.408477266032
RMSE: 12022.409082813892
R2:   0.04210200521035923


In [42]:
# Interpret the coefficient
print(f'Coefficient: {lr.coef_[0]:.2f}')
print(f'Intercept:   {lr.intercept_:.2f}')



Coefficient: 392.48
Intercept:   1292.81


Write your interpretation of the coefficient in this cell:
For every single point increase in BMI, predicted health insurance charges increase by $392.48, which is a really signficant increase


---

## Part 2.

Multiple Linear Regression with numeric features.

* Pick at least 3 of the numeric features in the dataset.
* Recreate your feature matrix, `X`
* Do another train test split.
* Make a Multiple Linear Regression.
* Evaluate your model.
* Interpret a coefficient (Remember the difference for interpreting coefficients of multiple linear regression vs. simple!)


In [43]:
# define a list of numeric features
numeric_features = ['bmi', 'age', 'children']

# define X, y
X = df[numeric_features]
y = df['charges']



In [44]:
# train test split
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)



In [45]:
# create a new instance of linear regression
mlr = LinearRegression()

# fit on the training data
mlr.fit(X_train, y_train)

# make predictions
y_pred = mlr.predict(X_test)


In [46]:
# Evaluate based on metrics
print('MAE: ', mean_absolute_error(y_test, y_pred))
print('RMSE:', root_mean_squared_error(y_test, y_pred))
print('R2:  ', r2_score(y_test, y_pred))



MAE:  9181.554834229108
RMSE: 11475.089985058188
R2:   0.12733318587824716


In [47]:
# interpret coefficients
for feature, coef in zip(numeric_features, mlr.coef_):
    print(f'{feature}: {coef:.2f}')
print(f'\nHolding all other features constant:')
print(f'  A 1-unit increase in BMI is associated with a ${mlr.coef_[0]:.2f} change in charges.')
print(f'  A 1-year increase in age is associated with a ${mlr.coef_[1]:.2f} change in charges.')
print(f'  One additional child is associated with a ${mlr.coef_[2]:.2f} change in charges.')



bmi: 332.38
age: 232.94
children: 654.80

Holding all other features constant:
  A 1-unit increase in BMI is associated with a $332.38 change in charges.
  A 1-year increase in age is associated with a $232.94 change in charges.
  One additional child is associated with a $654.80 change in charges.


Write your interpretation for your coefficient

*Hint:  remember the distinction between single and multiple linear regression when interpreting*
 BMI increase has a stronger potential increase in medical costs than age does, however having even one child is nearly double the relationship in increasing medical costs as the rest of the numerical factors




---

## Part 3.

### Add Categorical Variables.

* Select numeric and categorical variables for your feature matrix
* Encode the categorical variables using `pd.get_dummies()`
  * REMEMBER: There is some arguments you need to adjust inside of `pd.get_dummies()`
* Do a train test split
* Train and evaluate the model
* BONUS: Interpret a coefficient from one of the encoded categorical variables.

In [48]:
# Define features, X, y
features = ['age', 'bmi', 'children', 'sex', 'smoker', 'region']
X = df[features]
y = df['charges']



In [49]:
# Encode your categorical variables.
X = pd.get_dummies(X, drop_first=True)
X.head()



,age,bmi,children,sex_male,smoker_yes,region_northwest,region_southeast,region_southwest
0,19,27.900,0,False,True,False,False,True
1,18,33.770,1,True,False,False,True,False
2,28,33.000,3,True,False,False,True,False
3,33,22.705,0,True,False,True,False,False
4,32,28.880,0,True,False,True,False,False


In [50]:
# Do a train test split
X_train, X_test, y_train, y_test = train_test_split(X, y, random_state=42)

# instantiate a new model
model = LinearRegression()

# fit the model
model.fit(X_train, y_train)

# make predictions
y_pred = model.predict(X_test)


 

In [51]:
# Evalueate it
print('MAE: ', mean_absolute_error(y_test, y_pred))
print('RMSE:', root_mean_squared_error(y_test, y_pred))
print('R2:  ', r2_score(y_test, y_pred))


MAE:  4243.654116653138
RMSE: 5926.023602394469
R2:   0.7672642952734356


---

## Part 4.

Conceptual Wrap Up Questions:

* Why do we do a train, test split? We need to test out if the model works well on new and unseen data, not just the data it has learned from otherwise it is not going to be effective forward looking view. It is the difference if it is just spitting out the training data vs interpret and work on new data.
  
* Why do we have to handle our categorical variables in a particular way? We need to be specific in using numbers to translate the categorical variables so the model does not take into account a ranking or view of the factors (i.e. translating a binary yes no until 0 1 does not imply the 1 is superior to 0 or that we could rank/create mean or median of those results.)
  
* How did your model perform after adding more features? It significantly improved. The MAE over cut in half, so avg prediction error got a lost smaller. RMSE also cut in half, so overall prediction errors decreased. R2 also increased by nearly 5x, so the new method fits more of the variation in the data.
  
---

## Bonus Section: Comparing Training and Test Performance

So far, we have evaluated how well our model performs on the test data.

We can also check how well the model performs on the training data. This helps us compare:

- How well the model performs on data it learned from
- How well the model performs on new, unseen data

In `sklearn`, regression models have a `.score()` method.

For linear regression, `.score()` returns the model's **R² score**.

R² measures how much of the variation in the target variable is explained by the model.

Run the following code:

```python
print(f"Training score: {model.score(X_train, y_train):.3f}")
print(f"Testing score: {model.score(X_test, y_test):.3f}")
```

### How to Interpret the Scores

If the training score is much higher than the testing score, the model may be overfitting.

This means the model performs well on the data it learned from, but does not perform as well on new, unseen data.

If the training score and testing score are both low, the model may be underfitting.

This means the model is not learning enough of the relationship between the features and the target variable.

If the training score and testing score are close together, and both scores are reasonably strong for the problem, the model may be generalizing well.

This means the model is learning useful patterns from the training data and still performing well on new data.

### Reflection Question

Based on the training and testing scores, how would you characterize this model?

Would you say it is:

* Overfitting?
* Underfitting?
* Generalizing reasonably well?

Explain your answer in 2–3 sentences.

In [52]:
print(f"Training score: {model.score(X_train, y_train)}")
print(f"Testing score: {model.score(X_test, y_test)}")

Training score: 0.7449555328228536
Testing score: 0.7672642952734356


In [53]:
# this sounds like it is bonus and the two hw assignments + final project took its toll, so i will skip. if its mandatory ill go back and do it